src/client/SOAP_client.ts

```diff
@@ -1,22 +1,121 @@
-import * as path from 'path';
+import * as vscode from 'vscode';
+import { default as fetch }from 'node-fetch';
+import { createEnvelopeMessage, extractLogs, getResponseCode, MultipartMessage, ServerAction } from '../server/mime';
+import { ServerResult } from './server_result';
+import { log } from '../logger';
 
+/*
+1. Gather all the information for the message from the text (start with hardcoded)
+3. Create the message
+4. Figure out address, port etc and send message
+5. Parse response to get ID
+6. Poll server (GET at /p/<responseID>) until done
+7. Send "response" message to server
+8. Unpack response to get validation results
+*/
 
 export class SOAPClient {
-    url: string;
-    port: string;
-    url_dir: string;
-    method: string;
+    url: string = "http://build-oracc.museum.upenn.edu";
+    port: number = 8085;
     atf_filename: string;
     atf_text: string;
-    atf_path: string;
+    atf_project: string;
+    responseID: string;
 
-    constructor(atf_path: string, atf_text: string){
+    constructor(atf_name: string, atf_project: string, atf_text: string) {
         this.atf_text = atf_text;
-        this.port = "8085";
-        this.url = "http://build-oracc.museum.upenn.edu"
-        this.method = "POST";
-        this.atf_path = atf_path;
-        this.atf_filename = path.basename(atf_path);
+        this.atf_filename = atf_name;
+        this.atf_project = atf_project;
+        this.atf_text = atf_text;
+    }
+
+    async executeCommand(command: ServerAction): Promise<ServerResult> {
+        try {
+            this.responseID = await this.sendInitialRequest(command);
+            log('debug', `Got response code ${this.responseID}`);
+            await this.serverComplete();
+            log('info', `Request ${this.responseID} is done.`);
+            const finalResult = await this.retrieveReponse();
+            return new ServerResult(finalResult.get('oracc.log'),
+                                    finalResult.get('request.log'));
+        } catch (err) {
+            log('error', `Error during communication with server: ${err}`);
+            vscode.window.showErrorMessage(
+                "Problem getting response from server, see log for details.");
+            return;  // Should return an empty result?
```

Comment:
No need for an empty result to be returned I think. You're catching the errors and the ServerResults are returned via a Promise (which I don't quite understand yet!). What would you use an empty result for?

Reply:
The Promise represents a result that's not finished computing yet. At some point we'll need to either return a result, or say that there's an error. Maybe the right thing is to throw an error here, and let the caller show the message etc. I have trouble explaining it because I don't fully understand promises either.